test/e2e/storage/testsuites/multivolume.go

```diff
@@ -503,3 +527,62 @@ func TestConcurrentAccessToSingleVolume(f *framework.Framework, cs clientset.Int
 		utils.CheckReadFromPath(f, pod, *pvc.Spec.VolumeMode, path, byteLen, seed)
 	}
 }
+
+// getCurrentTopologies() goes through all Nodes and returns unique driver topologies and count of Nodes per topology
+func getCurrentTopologiesNumber(cs clientset.Interface, keys []string) ([]topology, []int, error) {
+	nodes, err := e2enode.GetReadySchedulableNodes(cs)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	topos := []topology{}
+	topoCount := []int{}
+
+	// TODO: scale?
+	for _, n := range nodes.Items {
+		topo := map[string]string{}
+		for _, k := range keys {
+			v, ok := n.Labels[k]
+			if !ok {
+				return nil, nil, fmt.Errorf("node %v missing topology label %v", n.Name, k)
+			}
+			topo[k] = v
+		}
+
+		found := false
+		for i, existingTopo := range topos {
+			if topologyEqual(existingTopo, topo) {
+				found = true
+				topoCount[i]++
+				break
+			}
+		}
+		if !found {
+			framework.Logf("found topology %v", topo)
+			topos = append(topos, topo)
+			topoCount = append(topoCount, 1)
+		}
+	}
+	return topos, topoCount, nil
+}
+
+// ensureTopologyRequirements sets nodeSelection affinity according to given topology keys for drivers that provide them
+func ensureTopologyRequirements(nodeSelection *e2epod.NodeSelection, cs clientset.Interface, topologyKeys []string, minCount int) error {
+	topologyList, topologyCount, err := getCurrentTopologiesNumber(cs, topologyKeys)
+	if err != nil {
+		return err
+	}
+	suitableTopologies := []topology{}
+	for i, topo := range topologyList {
+		if topologyCount[i] >= minCount {
+			suitableTopologies = append(suitableTopologies, topo)
+		}
+	}
+	if len(suitableTopologies) == 0 {
+		framework.Skipf("No topology with at least %d nodes found - skipping", minCount)
+	}
+	// Take the first suitable topology
+	e2epod.SetNodeAffinityTopologyRequirement(nodeSelection, suitableTopologies[0])
```

Comment:
ClientNode is already set in nodeSelection. Do we end up overwriting it?

Reply:
The function should add the required topology labels to the existing nodeSelection so not overwriting any existing fields there. Moreover the ClienNode seems to always be an empty string.

Reply:
ClientNode may be set by [external](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/external/external.go#L111) users. I guess they shouldn't also enable topology the same time. Otherwise, the two may conflict and you could end up with an unschedulable pod?

Reply:
Yes. I meant that in the case the ClientNode is really set the test would be skipped (every multi-node test checks this) so we'd never get here.